### PR TITLE
Alien acid now takes 3 seconds to apply on nuclear bombs with bonus sanity code.

### PR DESCRIPTION
**Pull request by Probe1**
Resolves https://github.com/vgstation-coders/vgstation13/issues/17438
Celt helped me do most of it so it's his fault if it's broken.  Not that it is I tested it to death.  Thank you for the help Celt.

### CODE REVIEW

File: code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm [hunk 1]
```diff
@@ -233,21 +242,13 @@ Doesn't work on other aliens/AI.*/
...
+			AdjustPlasma(-200)
 
 /proc/acidify(atom/O, mob/user)
-	if(O.acidable())
-		new /obj/effect/alien/acid(get_turf(O), O)
-		user.visible_message("<span class='alien'>\The [usr] vomits globs of vile stuff all over [O]. It begins to sizzle and melt under the bubbling mess of acid!</span>")
-	else
-		to_chat(user, "<span class='alien'>You cannot dissolve this object.</span>")
+	new /obj/effect/alien/acid(get_turf(O), O)
+	user.visible_message("<span class='alien'>\The [usr] vomits globs of vile stuff all over [O]. It begins to sizzle and melt under the bubbling mess of acid!</span>")
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: `usr` vs `user`

File: code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm [hunk 2]
```diff
@@ -214,8 +214,17 @@ Doesn't work on other aliens/AI.*/
...
 /spell/alienacid/is_valid_target(var/atom/target, mob/user)
-	if(get_dist(user, target) > range) //Shouldn't be necessary but a good check in case of overrides
+	return is_valid_target_to_acid(target,user,range)
+
+/proc/is_valid_target_to_acid(var/atom/target, mob/user,var/range=1)
+	if(get_dist(user, target) > range) 
+		to_chat(usr, "<span class='alien'>Target is too far away!</span>")
 		return FALSE
+	if(target.isacidhardened())
+		if(!do_after(usr,target,3 SECONDS))
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: `usr` vs `user`

File: code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm [hunk 3]
```diff
@@ -214,8 +214,17 @@ Doesn't work on other aliens/AI.*/
...
+			to_chat(usr, "<span class='alien'>You have to stay next to the object to acid it!</span>")
```
Comment (DamianX) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: `usr` vs `user`

**Comment by Probe1** [comment]
Well I knew it was a bad idea to use githubs built in conflict resolver.  But there I went pressing buttons.

File: code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm [hunk 4]
```diff
@@ -233,20 +242,16 @@ Doesn't work on other aliens/AI.*/
...
-				return FALSE
-			AdjustPlasma(-200)
+		if(is_valid_target_to_acid(O,usr))
 			acidify(O, usr)
-		else
-			to_chat(usr, "<span class='alien'>Target is too far away.</span>")
+			AdjustPlasma(-200)
 
 /proc/acidify(atom/O, mob/user)
 	if(O.acidable())
```
Comment (ghost) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: Reeeee remove this line why did you add it back

File: code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm [hunk 5]
```diff
@@ -233,20 +242,16 @@ Doesn't work on other aliens/AI.*/
...
-		new /obj/effect/alien/acid(get_turf(O), O)
-		user.visible_message("<span class='alien'>\The [user] vomits globs of vile stuff all over [O]. It begins to sizzle and melt under the bubbling mess of acid!</span>")
-	else
+    new /obj/effect/alien/acid(get_turf(O), O)
+	  user.visible_message("<span class='alien'>\The [user] vomits globs of vile stuff all over [O]. It begins to sizzle and melt under the bubbling mess of acid!</span>")
+  
+  else
```
Comment (ghost) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: And this line rrrreee

File: code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm [hunk 6]
```diff
@@ -233,20 +242,16 @@ Doesn't work on other aliens/AI.*/
...
 		to_chat(user, "<span class='alien'>You cannot dissolve this object.</span>")
```
Comment (ghost) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: And also this one gosh darn it

**Review by ghost — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm [hunk 7]
```diff
@@ -233,21 +242,13 @@ Doesn't work on other aliens/AI.*/
 		return
 
 	if(powerc(200))
-		if(O in oview(1))
-			if(!O.acidable())
-				to_chat(usr, "<span class='alien'>You cannot dissolve this object.</span>")
-				return FALSE
-			AdjustPlasma(-200)
+		if(is_valid_target_to_acid(O,usr))
```
Comment (ghost) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: Usr

File: code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm [hunk 8]
```diff
@@ -265,7 +266,7 @@ Doesn't work on other aliens/AI.*/
 /spell/aoe_turf/alienregurgitate/cast(list/targets, mob/user)
 	var/mob/living/carbon/alien/humanoid/alien = user
 	alien.drop_stomach_contents()
-	user.visible_message("<span class='alien'>\The [user] hurls out the contents of their stomach!</span>")
+	user.visible_message("<span class='alien'>\The [usr] hurls out the contents of their stomach!</span>")
```
Comment (ghost) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: Usr

File: code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm [hunk 9]
```diff
@@ -327,7 +328,7 @@ a
 	return ..()
 
 /spell/aoe_turf/evolve/drone/spell_do_after(var/mob/user as mob, delay as num, var/numticks = 5)
-	user.visible_message("<span class='alien'>[user] begins to violently twist and contort!</span>", "<span class='alien'>You begin to evolve, stand still for a few moments</span>")
+	user.visible_message("<span class='alien'>[src] begins to violently twist and contort!</span>", "<span class='alien'>You begin to evolve, stand still for a few moments</span>")
```
Comment (ghost) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: Should be user, not src

**Review by ghost — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm [hunk 10]
```diff
@@ -149,7 +149,7 @@ Doesn't work on other aliens/AI.*/
 	proj_step_delay = 0.2
 
 /spell/targeted/projectile/alienneurotoxin/is_valid_target(var/target, mob/user)
-	if(!(spell_flags & INCLUDEUSER) && target == user)
+	if(!(spell_flags & INCLUDEUSER) && target == usr)
```
Comment (ghost) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: User

File: code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm [hunk 11]
```diff
@@ -233,21 +242,13 @@ Doesn't work on other aliens/AI.*/
 		return
 
 	if(powerc(200))
-		if(O in oview(1))
-			if(!O.acidable())
-				to_chat(usr, "<span class='alien'>You cannot dissolve this object.</span>")
-				return FALSE
-			AdjustPlasma(-200)
+		if(is_valid_target_to_acid(O,user))
```
Comment (ghost) [inline code/modules/mob/living/carbon/alien/humanoid/alien_powers.dm]: Actually I lied this one is meant to be usr

**Comment by ghost** [comment]
Looks good 👌

**Comment by Probe1** [comment]
do_after is running twice for the spell and I don't know why.

**Comment by DamianX** [comment]
Looks like `is_valid_target()` is called by both `channeled_spell()` and `before_cast()`.
https://github.com/vgstation-coders/vgstation13/blob/7bdd3f77cf82e47139ab330a7112b55bd57a569e/code/modules/spells/spell_code.dm#L213-L214

**Pull request merged**